### PR TITLE
Remember state for console logs, improvements to pages whose state should be restored after navigating back

### LINKED ISSUE #1261: Console logs: Remember last selected resource
I added logic to the metrics page to remember the last selected resource. If you navigate away and come back, it is automatically re-selected. We should do the same thing on the console logs page.

console-logs-remember

Metrics page saves the selected state in `ProtectedSessionStorage`.

### LINKED ISSUE #1446: Navigating to metrics page will take you to last metrics page state even if you're not going to the base path
ie: A user navigates to /Metrics/xyz and then away from that page. If they go back to /Metrics, they will be redirected to /Metrics/xyz. This is correct.

A user navigates to /Metrics/xyz and then away from that page. But if they attempt to go directly to /Metrics/abc, they will still be redirected to /Metrics/xyz. This is incorrect, we should only navigate back to state stored in local storage if the user is navigating to the base path.

### PR DESCRIPTION
**Pull request by adamint**
Adds `IPageWithSessionAndUrlState`, which handles the following tasks:
- If the user navigates back to the page, 1) restore page state from the session, if available, or 2) compute state from the query (`InitializeViewModelAsync`)
- If user navigates to a specific subpage/with specific data, do not attempt to restore page state (fixes #1446) 
- Requires consumers to put page state in a view model, to avoid bloating the component (`ViewModel`)
- After a change is made to the view model, 1) navigates to the new page url, and 2) saves the new state in the session (`AfterViewModelChangedAsync`)

Because the view model may include specific objects that we don't want to or cannot serialize, `ConvertViewModelToSerializable` converts `TViewModel` to `TSerializableViewModel`.

The rest of the PR is rewriting `Metrics` to implement this new interface, as well as `ConsoleLogs` (fixes #1261). I tried to make this interface flexible to be fairly easily implementable in additional pages, if we want to remember state in additional pages. Please let me know if I'm overly complicating it!
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1453)

### CODE REVIEW

**Comment by JamesNK** [comment]
I like the idea of putting structure around this pattern rather than each page duplicating logic and doing it slightly differently. That's on the right track. The thing that gives me pause is making it an interface on the page. Maintain-ability of composition scales better than inheritance/implementation as it keeps things separated.

To give good feedback I'll need to get your branch and play around with it. My feedback might come in a branch with some changes.

**Comment by adamint** [comment]
I'm not sure composition is a great fit in this instance, given how coupled the `state manager` and `page` parent objects would be, but I'm happy to hear what you all think, and any design ideas here. 

@drewnoakes @tlmii do you have thoughts?

**Comment by adamint** [comment]
Could I get a review on this, please

**Comment by JamesNK** [comment]
I went ahead and made changes myself in a branch. I wanted to see how it worked end-to-end and made changes while testing it.

My changes are here: https://github.com/dotnet/aspire/commit/8a8e5fa4889f64f8c913605cfc964d5e75fc3412. I think everything in it is good. You could cherry-pick the changes into your branch.

File: src/Aspire.Dashboard/Components/Pages/IPageWithSessionAndUrlState.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+        page.NavigationManager.NavigateTo(pathWithParameters);
+        await page.SessionStorage.SetAsync(page.SessionStorageKey, serializableViewModel).ConfigureAwait(false);
+    }
+
+    /// <summary>
+    /// Called to initialize the page's view model. If <see cref="hasComponentRendered"/> is true,
+    /// will try to restore from session storage. Otherwise, it will construct state from the url.
+    ///
+    /// Because session storage isn't accessible until after component render, this should be called twice:
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Pages/IPageWithSessionAndUrlState.cs:73]: The reason you couldn't access session storage before render is accessing it forces a component render. The view model is still null at this point, so there is an error when render tries to use the null view model.

I removed the requirement to call `InitializeViewModelAsync` twice by creating the view model in initialize. `GetViewModelFromQuery` changes to an update.

Comment (adamint) [reply src/Aspire.Dashboard/Components/Pages/IPageWithSessionAndUrlState.cs]: Ah, thank you for figuring that out 😄

**Comment by adamint** [comment]
Thanks @JamesNK 😄 your changes are now in the branch.

**Review by JamesNK — APPROVED** [review]

**Pull request merged**